resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -308,114 +309,88 @@ private[spark] class Client(
-                sparkConf.set(KUBERNETES_DRIVER_SERVICE_NAME, service.getMetadata.getName)
-                sparkConf.set("spark.app.id", kubernetesAppId)
-                sparkConf.setIfMissing("spark.app.name", appName)
-                sparkConf.setIfMissing("spark.driver.port", DEFAULT_DRIVER_PORT.toString)
-                sparkConf.setIfMissing("spark.blockmanager.port",
-                  DEFAULT_BLOCKMANAGER_PORT.toString)
-                val driverSubmitter = buildDriverSubmissionClient(kubernetesClient, service,
+                val ownerRefs = Seq(new OwnerReferenceBuilder()
+                  .withName(pod.getMetadata.getName)
+                  .withUid(pod.getMetadata.getUid)
+                  .withApiVersion(pod.getApiVersion)
+                  .withKind(pod.getKind)
+                  .withController(true)
+                  .build())
+                applicationSecrets.foreach(secret => {
+                  secret.getMetadata.setOwnerReferences(ownerRefs.asJava)
+                  kubernetesClient.secrets().createOrReplace(secret)
+                })
+                val driverSubmissionServicePort = new ServicePortBuilder()
+                  .withName(SUBMISSION_SERVER_PORT_NAME)
+                  .withPort(SUBMISSION_SERVER_PORT)
+                  .withNewTargetPort(SUBMISSION_SERVER_PORT)
+                  .build()
+                val service = kubernetesClient.services().createNew()
+                  .withNewMetadata()
+                    .withName(kubernetesAppId)
+                    .withLabels(driverKubernetesSelectors)
+                    .withOwnerReferences(ownerRefs.asJava)
+                    .endMetadata()
+                  .withNewSpec()
+                    .withType("NodePort")
+                    .withSelector(driverKubernetesSelectors)
+                    .withPorts(driverSubmissionServicePort)
+                    .endSpec()
+                  .done()
+
+                try {
+                  sparkConf.getOption("spark.app.id").foreach { id =>
+                    logWarning(s"Warning: Provided app id in spark.app.id as $id will be" +
+                      s" overridden as $kubernetesAppId")
+                  }
+                  sparkConf.set(KUBERNETES_DRIVER_POD_NAME, kubernetesAppId)
+                  sparkConf.set(KUBERNETES_DRIVER_SERVICE_NAME, service.getMetadata.getName)
+                  sparkConf.set("spark.app.id", kubernetesAppId)
+                  sparkConf.setIfMissing("spark.app.name", appName)
+                  sparkConf.setIfMissing("spark.driver.port", DEFAULT_DRIVER_PORT.toString)
+                  sparkConf.setIfMissing("spark.blockmanager.port",
+                    DEFAULT_BLOCKMANAGER_PORT.toString)
+                  val driverSubmitter = buildDriverSubmissionClient(kubernetesClient, service,
                     driverSubmitSslOptions)
-                val ping = Retry.retry(5, 5.seconds) {
                   driverSubmitter.ping()
```

Comment:
comment that this ping is a final check of the service liveness before submitting the job (in addition to the k8s checks).

Might also be worth try/catching it and logging on failure that even though the k8s service is active we are unable to connect to the driver's rest service from the submitter